### PR TITLE
Putting debug info in PDB container

### PR DESCRIPTION
**Pull request by adam-yang**

### CODE REVIEW

File: lib/DXIL/DxilPDB.cpp [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+
+
+struct PDBReader {
+  IStream *m_pStream = nullptr;
+  IMalloc *m_pMalloc = nullptr;
+  UINT32 m_uOriginalOffset = 0;
+  MSF_SuperBlock m_SB = {};
+  HRESULT m_Status = S_OK;
+
+  HRESULT SetPostion(INT32 sOffset) {
```
Comment (MrTrillian) [inline lib/DXIL/DxilPDB.cpp]: (typo)

File: lib/DXIL/DxilPDB.cpp [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+    if (memcmp(pSB->MagicBytes, kMsfMagic, sizeof(kMsfMagic)) != 0)
+      return E_FAIL;
+
+    return S_OK;
+  }
+
+  HRESULT ReadU32(UINT32 *pValue) {
+    ULONG NumBytesRead = 0;
+    support::ulittle32_t ValueLE;
+    IFR(m_pStream->Read(&ValueLE, sizeof(ValueLE), &NumBytesRead));
```
Comment (MrTrillian) [inline lib/DXIL/DxilPDB.cpp]: Is it a concern that we always ignore the `numbytesread`? We could have a ReadAllBytes(IStream, void*, SIZE_T) that fails if the stream reaches its end and does not read everything.

File: lib/DxilDia/DxilDiaDataSource.cpp [hunk 3]
```diff
@@ -65,11 +66,25 @@ std::unique_ptr<llvm::MemoryBuffer> getMemBufferFromStream(_In_ IStream *pStream
...
+      return E_FAIL;
+    CComPtr<IDxcBlobEncoding> pPinnedBlob;
+    IFR(hlsl::DxcCreateBlobWithEncodingFromPinned(PartHeader+1, PartHeader->PartSize, CP_UTF8, &pPinnedBlob));
+    pIStream.Release();
+    IFR(hlsl::CreateReadOnlyBlobStream(pPinnedBlob, &pIStream));
+  }
+
   m_context.reset();
   m_finder.reset();
   try {
```
Comment (MrTrillian) [inline lib/DxilDia/DxilDiaDataSource.cpp]: Exceptions can happen before this point as the methods above use collections that might throw them.

File: lib/DxilDia/DxilDiaDataSource.cpp [hunk 4]
```diff
@@ -65,11 +66,25 @@ std::unique_ptr<llvm::MemoryBuffer> getMemBufferFromStream(_In_ IStream *pStream
...
+
+  CComPtr<IStream> pIStream = pInputIStream;
+  CComPtr<IDxcBlob> pContainer;
+  if (SUCCEEDED(hlsl::pdb::LoadDataFromStream(m_pMalloc, pInputIStream, &pContainer))) {
+    hlsl::DxilPartHeader *PartHeader =
+      hlsl::GetDxilPartByType((hlsl::DxilContainerHeader *)pContainer->GetBufferPointer(), hlsl::DFCC_ShaderDebugInfoDXIL);
+    if (!PartHeader)
+      return E_FAIL;
+    CComPtr<IDxcBlobEncoding> pPinnedBlob;
+    IFR(hlsl::DxcCreateBlobWithEncodingFromPinned(PartHeader+1, PartHeader->PartSize, CP_UTF8, &pPinnedBlob));
```
[the remark this answers is not in the document]
Comment (adam-yang) [reply lib/DxilDia/DxilDiaDataSource.cpp]: Fixed it.

File: lib/DXIL/DxilPDB.cpp (hunk 1 above)
Comment (adam-yang) [reply lib/DXIL/DxilPDB.cpp]: Fixed.

File: lib/DXIL/DxilPDB.cpp (hunk 2 above)
Comment (adam-yang) [reply lib/DXIL/DxilPDB.cpp]: Good idea.

**Pull request merged**